### PR TITLE
Fix foreach deconstruction nullability

### PR DESCRIPTION
**Pull request by YairHalberstadt**
Use the flow state to set the nullability of deconstruction variables in foreach when deconstructing a tuple.

Fixes https://github.com/dotnet/roslyn/issues/39736, https://github.com/dotnet/roslyn/issues/52042, and https://github.com/dotnet/roslyn/issues/37491.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 1]
```diff
@@ -80252,8 +80252,11 @@ void M()
             // from the deconstruction case
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:80252]: Comment should be removed as well.

**Review by 333fred — COMMENTED** [review]
Sorry for the long delay on reviews Yair. This looks overall fine to me, but has another test that needs updating and the bootstrap build needs to be fixed.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 2]
```diff
@@ -80356,8 +80356,11 @@ void M()
...
```
Comment (YairHalberstadt) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:80356]:
```suggestion

```

Comment (YairHalberstadt) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:80356]:
```suggestion

```

**Comment by 333fred** [comment]
@YairHalberstadt looks like there are some compile and test errors (compile error might be the bootstrap compiler with this change finding a potential issue).

**Comment by RikkiGibson** [comment]
Hi @YairHalberstadt, your change introduced a new warning to the compiler:

`src\Workspaces\Core\Portable\Editing\ImportAdderService.cs(135,43): error CS8604: (NETCORE_ENGINEERING_TELEMETRY=Build) Possible null reference argument for parameter 'item' in 'bool HashSet<INamespaceSymbol>.Contains(INamespaceSymbol item)'.`

https://github.com/dotnet/roslyn/blob/6543c7086f9154477330b01bb43f45d897f36325/src/Workspaces/Core/Portable/Editing/ImportAdderService.cs#L124-L136

I think solution here is probably to add a `!` to the usage of `namespaceSymbol`, or to eliminate the `Where` and insert an `if` into the loop. Up to you :)

Please let us know if there's any way we can provide assistance with the PR.

**Comment by YairHalberstadt** [comment]
Hey @RikkiGibson 

So I've moved over to Linux recently and I haven't yet got round to setting up my machine so it can build Roslyn. I would like to but I'm pretty busy at the moment. So feel free to work off this PR if it's useful for you, or I'll eventually get round to it, but maybe not this month. Feel free to close and I can reopen if that's helpful.

Thanks for your understanding!

**Comment by RikkiGibson** [comment]
Ping @dotnet/roslyn-compiler

**Comment by RikkiGibson** [comment]
Ping @dotnet/roslyn-ide for a nullable warning fix

**Review by 333fred — APPROVED** [review]

**Comment by RikkiGibson** [comment]
@CyrusNajmabadi @dotnet/roslyn-ide for sign-off on IDE changes (resolving nullability warnings).

File: src/Workspaces/Core/Portable/Editing/ImportAdderService.cs [hunk 3]
```diff
@@ -121,12 +121,14 @@ private static SyntaxNode GenerateNamespaceImportDeclaration(INamespaceSymbol na
...
 
             var nodesWithExplicitNamespaces = syntaxNodes
-                .Select(n => (syntaxnode: n, namespaceSymbol: GetExplicitNamespaceSymbol(n, model)))
-                .Where(x => x.namespaceSymbol != null);
+                .Select(n => (syntaxnode: n, namespaceSymbol: GetExplicitNamespaceSymbol(n, model)));
 
             var addedSymbols = new HashSet<INamespaceSymbol>();
             foreach (var (node, namespaceSymbol) in nodesWithExplicitNamespaces)
             {
+                if (namespaceSymbol is null)
```
Comment (RikkiGibson) [inline src/Workspaces/Core/Portable/Editing/ImportAdderService.cs:129]: We used to incorrectly assume that the 'namespaceSymbol' was not-null at the beginning of the loop. Now that we fixed nullable deconstruction analysis we have to null check it in a way the compiler understands.

File: src/Workspaces/Core/Portable/Editing/ImportAdderService.cs [hunk 4]
```diff
@@ -121,12 +121,14 @@ private static SyntaxNode GenerateNamespaceImportDeclaration(INamespaceSymbol na
             var model = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
 
             var nodesWithExplicitNamespaces = syntaxNodes
-                .Select(n => (syntaxnode: n, namespaceSymbol: GetExplicitNamespaceSymbol(n, model)))
-                .Where(x => x.namespaceSymbol != null);
```
Comment (jmarolf) [inline src/Workspaces/Core/Portable/Editing/ImportAdderService.cs:125]: Why was this removed? haven't we just moved to null check into the foreach loop now?

Comment (RikkiGibson) [reply src/Workspaces/Core/Portable/Editing/ImportAdderService.cs:125]: Yes. Moving the null check into the foreach is necessary for the compiler to realize that 'namespaceSymbol' is not null. Before the compiler just assumed 'namespaceSymbol' was not null. See https://github.com/dotnet/roslyn/pull/52722/files#r782643742

File: src/Features/Core/Portable/AddFileBanner/AbstractAddFileBannerCodeRefactoringProvider.cs [hunk 5]
```diff
@@ -78,7 +78,7 @@ public override async Task ComputeRefactoringsAsync(CodeRefactoringContext conte
             {
                 cancellationToken.ThrowIfCancellationRequested();
 
-                var siblingBanner = await TryGetBannerAsync(siblingDocument, siblingRoot, cancellationToken).ConfigureAwait(false);
+                var siblingBanner = await TryGetBannerAsync(siblingDocument, siblingRoot!, cancellationToken).ConfigureAwait(false);
```
Comment (jmarolf) [inline src/Features/Core/Portable/AddFileBanner/AbstractAddFileBannerCodeRefactoringProvider.cs:81]: `TryGetBannerAsync` should be updated to have a signature of 
```csharp
Task<ImmutableArray<SyntaxTrivia>> TryGetBannerAsync(Document document, SyntaxNode? root, CancellationToken cancellationToken)
```

instead of suppressing this nullable warning. The implementation already checks if its null.

**Review by jmarolf — CHANGES_REQUESTED** [review]
please don't suppress the nullable warning when passing to the `TryGetBannerAsync` method

**Review by jmarolf — APPROVED** [review]

**Comment by RikkiGibson** [comment]
Thanks for the contribution @YairHalberstadt! (we finally got it merged 😄)
